### PR TITLE
client: introduce ExchangeFunc and Unwrap() to access [*dns.Client]

### PR DESCRIPTION
**Pull request by amery**

### CODE REVIEW

File: pkg/client/client.go [hunk 1]
```diff
@@ -10,13 +10,27 @@ import (
...
 	_ Client = (*dns.Client)(nil)
+	_ Client = (ClientFunc)(nil)
 )
 
 // A Client makes a request to a server
 type Client interface {
 	ExchangeContext(context.Context, *dns.Msg, string) (*dns.Msg, time.Duration, error)
 }
 
+// revive:disable:exported
```
Comment (karasz) [inline pkg/client/client.go:21]: I do not like this but I do not have a better alternative for now `resolver.Client.ClientFunc` is ugly, what about:
```go
type Client interface {
    ExchangeContext(context.Context, *dns.Msg, string) (*dns.Msg, time.Duration, error)
}

// ExchangeFunc is a function that implements the [Client]  interface
type ExchangeFunc func(context.Context, *dns.Msg, string) (*dns.Msg, time.Duration, error)
```
that would provide `resolver.Client.ExchangeFunc`

But I am not sold on an ExchangeFunc that implements the Client interface even if an interface can be comprised of multiple functions so an one to one correspondence between the interface name and the function name is impossible.

Comment (amery) [reply pkg/client/client.go]: amended accordingly

Comment (amery) [reply pkg/client/client.go]: as this lives within the `client` package I don't see a problem with `ExchangeFunc`

**Review by karasz — APPROVED** [review]
LGTM

**Comment by amery** [comment]
`v0.7.6` tagged

**Pull request merged**